### PR TITLE
chart-diagrams: Port to diagrams-1.3

### PR DESCRIPTION
**Pull request by bgamari**
This ports the `diagrams` backend to `diagrams-1.3` (as well as cleans up some style issues). I have yet to test with GHC 7.10 due to GHC bug [#10288](https://ghc.haskell.org/trac/ghc/ticket/10288#comment:12).

### CODE REVIEW

**Comment by bgamari** [comment]
This appears to work fine with the current state of the `ghc-7.10` branch and https://github.com/diagrams/diagrams-svg/pull/80 (which fixes a bug in `diagrams-svg`).

**Comment by bgamari** [comment]
Actually, it seems like the colors are off in some of the EPS output. Is this a known issue with previous releases?

**Comment by byorgey** [comment]
Released the bug fix for `diagrams-svg` to Hackage.

**Comment by timbod7** [comment]
See question on the related pull requests: https://github.com/timbod7/haskell-chart/pull/82

Which patch to choose?

**Comment by idontgetoutmuch** [comment]
What's the news on this and on #82?

**Comment by timbod7** [comment]
I've merged this with master, in an attempt to use the newly unified test harness with it. The results are in 

https://github.com/timbod7/haskell-chart/tree/diagrams-1.3

unfortunately, the diagrams related modes of the harness all freeze, eg:

```
harness drawing-diagrams
```

Have I done something silly?

**Comment by byorgey** [comment]
I could try to look into this.  Can you give me instructions on how to build & run the test harness?

**Comment by timbod7** [comment]
The test harness can be built with `cabal build` in the chart-tests directory.

The `make tests` target in the top level makefile captures the steps required to build the harness and generate all test output.

To see the problem you'll need to be on the diagrams-1.3 branch.

File: chart-diagrams/Graphics/Rendering/Chart/Backend/Diagrams.hs [hunk 1]
```diff
@@ -215,33 +220,33 @@ cBackendToEmbeddedFontSVG cb env = (svg, x)
...
         let font = envSelectFont env $ fs { _font_name = fFam
                                           , _font_slant = fSlant
-                                          , _font_weight = fWeight 
+                                          , _font_weight = fWeight
                                           }
         makeSvgFont font usedGs
         -- M.Map (String, FontSlant, FontWeight) (S.Set String)
         -- makeSvgFont :: (FontData, OutlineMap) -> Set.Set String -> S.Svg
-    svg = D.renderDia DSVG.SVG (DSVG.SVGOptions (D2.Dims w h) fontDefs) d
+    svg = D.renderDia DSVG.SVG (DSVG.SVGOptions (D2.dims2D w h) Nothing T.empty) d
```
Comment (byorgey) [inline chart-diagrams/Graphics/Rendering/Chart/Backend/Diagrams.hs]: What happened to the `fontDefs`?

**Comment by byorgey** [comment]
OK, thanks.  Building now.  I also took a careful look through all the diffs in this pull request and everything looks fairly straightforward.

**Comment by byorgey** [comment]
For me it does not hang, though it is rather slow (took several minutes to run `harness drawing-diagrams`.

**Comment by byorgey** [comment]
If that is much slower than it used to be then perhaps we should look into why it is so slow.

**Comment by timbod7** [comment]
It does indeed seem to be much slower.

master:

```
22:58:39 {master} ~/repos/chart$ cabal sandbox hc-pkg list | grep diagrams
    Chart-diagrams-1.4
    diagrams-cairo-1.2.0.7
    diagrams-core-1.2.0.6
    diagrams-lib-1.2.0.9
    diagrams-postscript-1.1.0.5
    diagrams-svg-1.1.0.5
22:59:04 {master} ~/repos/chart$ time ./.cabal-sandbox/bin/harness drawing-diagrams

real	0m5.970s
user	0m5.514s
sys	0m0.399s
22:59:33 {master} ~/repos/chart$ 
```

diagrams-1.3

```
22:58:42 {diagrams-1.3} ~/repos/chart-bgamari$ cabal sandbox hc-pkg list | grep diagrams
    Chart-diagrams-1.4.1
    diagrams-cairo-1.3.0.2
    diagrams-core-1.3.0.1
    diagrams-lib-1.3.0.1
    diagrams-postscript-1.3.0.1
    diagrams-solve-0.1
    diagrams-svg-1.3.1.2
22:59:39 {diagrams-1.3} ~/repos/chart-bgamari$ time ./.cabal-sandbox/bin/harness drawing-diagrams

real	3m16.784s
user	3m4.017s
sys	0m8.995s
23:03:07 {diagrams-1.3} ~/repos/chart-bgamari$
```

**Comment by byorgey** [comment]
Ouchies!  That is comically bad.  Let's see if we can figure out what is going on.  Later today I will see if I can do some profiling.

**Comment by bgamari** [comment]
Sorry for the latency! My laptop has had an unfortunate incident involving a glass of tea and consequently internet access will be a bit spotty for a few days. As I recall things were indeed quite slow.

**Comment by bergey** [comment]
It was easy to  narrow the problem down to `SVGFonts`.  `defaultFonts` in `chart-diagrams` calls `loadFont` in `SVGFonts`, which accounts for an impressive 99.9% of the runtime (and allocations) for `harness drawing-diagrams`.  I commented out most of `defaultFonts` to only load one font - this brings the runtime down by a factor of 20 (10m39s to 36s), but 98% of the time is still in `loadFont`.  I was hoping this change would give enough resolution to notice any other slow sections, but maybe it's just the one.

[Call tree for the unmodified test suite](http://bergey.s3.amazonaws.com/diagrams/harness.prof) (search for 99.9)
[Call tree with only one font loaded](http://bergey.s3.amazonaws.com/diagrams/2015-06-01-harness-only-SourceSansPro_R.prof) (search for 97.8)
(4 MB each)

Beyond that, I'm not sure what's changed.  I don't see any changes to font handling in `chart-diagrams` recently.  I've never really looked at the code for `SVGFonts`, so I hope @tkvogt, @jeffreyrosenbluth or someone else can look more closely.

**Comment by tkvogt** [comment]
I can maybe fix this. My plan was to make SVGFonts depend on diagrams-input, because the svg parser in diagrams-input  has a more correct path handling (d="M..") and is faster because of attoparsec and xml-conduit. Then half of the code in SVGFonts will vanish and maybe this bug will also go away.

**Comment by bgamari** [comment]
@tkvogt @bergey right, now this is coming back to me. I do remember tracing this back to `SVGFonts`. I have a branch moving the parsing over to `attoparsec` although don't recall it making a large difference. Unfortunately the branch is on the m2 drive of my dead laptop at the moment so it's not terribly accessible.

I think `SVGFonts` could be made faster although I also think that `Chart-diagrams` could be a bit more conservative about how many fonts it asks for. As it stands `defaultFonts` unconditionally loads ten fonts regardless of whether the diagram even has any text.

File: chart-diagrams/Graphics/Rendering/Chart/Backend/Diagrams.hs (hunk 1 above)
Comment (bgamari) [inline chart-diagrams/Graphics/Rendering/Chart/Backend/Diagrams.hs]: Ahem, right, I needlessly dropped it. Good catch.

**Comment by byorgey** [comment]
Is it possible that the call to `loadFont` is getting inlined somehow, so that the actual work of loading the font is happening many times?

**Comment by tkvogt** [comment]
@bgamari Did you replace xml-parsing or path parsing, or both in your branch?
The path parsing has to be replaced, because there is an open bug: https://github.com/diagrams/SVGFonts/issues/15
It's quite likely that the path parsing in diagrams-input is now fully correct.
I also think that the xml library of Galois is slower than xml-conduit because it goes trough the file (sub-) streams repeatedly while xml-conduit uses just one pass.

**Comment by bgamari** [comment]
@tkvogt only path parsing as far as I can recall.

**Comment by bgamari** [comment]
For the record, I've rebased against `master` and squashed some of the commits together.

**Comment by byorgey** [comment]
As a data point here, I instrumented `loadDataFont` in `Graphics.SVGFonts.Fonts` by changing the definition to
```
loadDataFont = trace "loadDataFont" (unsafePerformIO . loadFont . dataFile)
```
Now I can see that `loadDataFont` is being called repeatedly, in fact, once per call to `textSVG`!  I am not sure of the proper way to make sure that it only gets called once and the result cached.   I suspect that the type class polymorphism is not helping anything. The `unsafePerformIO` is not helping anything either.  It is a perfectly valid use of `unsafePerformIO` from a *semantic* point of view, but from an *operational* point of view it is problematic.

**Comment by byorgey** [comment]
Correction: `loadDataFont` is called repeatedly when compiling the final executable with no optimization turned on.  With `-O2`, `loadDataFont` only gets called once.

**Comment by byorgey** [comment]
I'm going to open an issue on the `SVGFonts` repo, we can keep discussing it there.

**Comment by tkvogt** [comment]
Could it be that simple:
The compiler can't recognize if a function is called again and again with the same argument.
One way to cache a result is to put it into a value without a parameter:
lin = loadDataFont "fonts/LinLibertine.svg"
We maybe have to use [memoization](https://wiki.haskell.org/Memoization).
Just my guess.

**Comment by byorgey** [comment]
Yes, but `lin = loadDataFont "fonts/LinLibertine.svg"` is not enough, because (1) it still has a typeclass polymorphic type, which means it may get re-run every time it is instantiated with a dictionary, and (2) it may get inlined.  (Incidentally, I did try putting a `NOINLINE` pragma on `lin` but it didn't seem to make any difference.)

**Comment by bgamari** [comment]
Any outcome here?

@tkvogt you'll find a pull request for the attoparsec rework above.

**Comment by tkvogt** [comment]
I am in the middle of merging SVGFonts into the svg parser of diagrams-input, beacuse they logically belong together. But I don't want to let you wait for this. As mentioned [here](https://github.com/diagrams/SVGFonts/issues/16#) I will try a global lookup, test it and then upload it to hackage.

**Comment by bgamari** [comment]
@timbod7 what do you think about just merging this? As far as I know the only issues are performance problems that can be dealt with later here or in SVGFonts.

**Comment by timbod7** [comment]
I want to see this in as much as anyone...

However, It's 2 orders of magnitude slower than the existing diagrams backend, which is already significantly slower than the cairo backend.  Even the tests don't finish in a reasonable about of time.

**Comment by bgamari** [comment]
Fair enough.

**Comment by timbod7** [comment]
Not wanting to completely sideline the good work done on this so far, I spend a couple of hours tinkering this evening. There definitely seems to have been some performance regression in SVGFonts. This is exacerbated by the fact that one test harness inadvertently loads the fonts multiple times. I fixed that harness, and  I've made a couple of API breaking changes to further reduce the impact of font loading:

1) By default now only sans-serif fonts are loaded (ie 6 faces are loaded instead of 14).
2) There is now an API to let the caller preload all fonts, giving explicit control.

These go some way to addressing the immediate problem, but I'm still unsure about what a pure API for font selection and text rendering should look like.

Given the API change, the version will need to bump to 1.5 before a release.

**Pull request merged**